File: cpp/all-your-base/all_your_base.cpp
#include "all_your_base.h"

namespace all_your_base {

std::vector<unsigned int>
convert(const unsigned int input_base,
        const std::vector<unsigned int> input_sequence,
        const unsigned int output_base) {
    if (input_base <= 1) {
        throw std::invalid_argument{
            "input base must be greaterthan or equal to 2"};
    }

    if (output_base <= 1) {
        throw std::invalid_argument{
            "output base must be greaterthan or equal to 2"};
    }

    if (input_sequence.empty()) {
        return std::vector<unsigned int>{};
    }

    if (*std::max_element(input_sequence.begin(), input_sequence.end()) >=
        input_base) {
        throw std::invalid_argument{
            "invalid input sequence for given input base"};
    }

    auto base2ten_op = [input_base](unsigned int base_ten, unsigned int digit) {
        base_ten *= input_base;
        base_ten += digit;

        return base_ten;
    };

    unsigned int base_ten{std::accumulate(
        input_sequence.begin(), input_sequence.end(), 0U, base2ten_op)};

    std::vector<unsigned int> output_sequence{};

    while (base_ten > 0) {
        output_sequence.push_back(base_ten % output_base);

        base_ten /= output_base;
    }

    std::reverse(output_sequence.begin(), output_sequence.end());

    return output_sequence;
}

} // namespace all_your_base

File: lua/all-your-base/all-your-base.lua
local all_your_base = {}

local function to_base_10(input_base, input_digits)
	local base_ten = 0

	for _, digit in ipairs(input_digits) do
		assert(digit >= 0, "negative digits are not allowed")
		assert(digit < input_base, "digit out of range")

		base_ten = base_ten * input_base
		base_ten = base_ten + digit
	end

	return base_ten
end

local function from_base_10(output_base, number)
	local output_digits = {}

	while number > 0 do
		local tmp = math.floor(number % output_base)

		table.insert(output_digits, 1, tmp)

		number = math.floor(number / output_base)
	end

	if #output_digits == 0 then
		return { 0 }
	else
		return output_digits
	end
end

all_your_base.convert = function(from_digits, from_base)
	assert(from_base > 1, "invalid input base")

	local function to(to_base)
		assert(to_base > 1, "invalid output base")

		return from_base_10(to_base, to_base_10(from_base, from_digits))
	end

	return { to = to }
end

return all_your_base
